cpp/src/arrow/util/bit_util.h

```diff
@@ -167,10 +167,8 @@ constexpr int64_t CoveringBytes(int64_t offset, int64_t length) {
 
 // Returns the 'num_bits' least-significant bits of 'v'.
 static inline uint64_t TrailingBits(uint64_t v, int num_bits) {
-  if (ARROW_PREDICT_FALSE(num_bits == 0)) return 0;
   if (ARROW_PREDICT_FALSE(num_bits >= 64)) return v;
-  int n = 64 - num_bits;
-  return (v << n) >> n;
+  return ((v >> num_bits) << num_bits) ^ v;
```

Comment:
What about `return v & ~(-1ULL << num_bits);` ?
It enables gcc to optimize the code with bmi2 `bzhi`.
https://godbolt.org/z/oq9zx4nhf
And looks it's slightly faster even without bmi2.
https://quick-bench.com/q/rgBQzUFls9IP48xm3JiRPtJoI_M

Reply:
Aha I've tried on clang 17.0.1 and it doesn't generate `bzhi`, compiler is so hacking...